### DISCUSSION TITLE
Support for non-Node runtimes (e.g., browsers, edges)

### DISCUSSION THREAD

**Discussion by ciscorn** [svelte-docgen/svelte-docgen#47] [Ideas]
Currently, `svelte-docgen` is designed for Node-compatible environments. However, browsers and major edge environments (e.g., Cloudflare Workers, Vercel’s edge runtime, etc.) rely on web-standard APIs, making these Node-specific dependencies a limiting factor.

To make `svelte-docgen` versatile, we should minimize its reliance on Node-specific modules wherever possible.

The TypeScript compiler, which is our main backend, has a well-abstracted system interface and can work in non-Node environments.

**Comment by xeho91** [comment]
I suspect this is doable, given your recent exploration.

My only concern is how to deal without using I/O operations _(e.g., with `node:fs`)_?
For example, to determine the workspace project root _(monorepo)_, like this one function:
https://github.com/svelte-docgen/svelte-docgen/blob/49d1a9d7cd74d1889cf80ea9df2115b3b99f2c8d/packages/svelte-docgen/src/shared.js#L172-L201

**Reply by xeho91** [reply]
I've just noticed that the function above is the _only_ one that uses `node:*` modules in two packages:

- `@svelte-docgen/extractor`
- `svelte-docgen`

I'll try to see what else I can do get the project root path url without using those `node:*` modules.

**Reply by xeho91** [reply]
One solution could be enforcing users to provide `Options.root`, so we don't need to determine it for him.

But... I'm not a fan of it. In the end, we only need this function to prevent exposing the full path to third parties in `sources` set entries.
Perhaps I should think about different ways of trimming down?

TypeScript libs types will be located in `node_modules/`.
Svelte types as well.
So, perhaps we just trim down anything before the first `<to trim>/node_modules/**/*`...

However, this doesn't apply to extending interfaces/props _inside_ the project. I'm stuck on this part.

**Reply by xeho91** [reply]
Silly me, I should read better the last sentence:

> The TypeScript compiler, **which is our main backend, has a well-abstracted system** interface and can work in non-Node environment

This TypeSccript `ts.System` _provides_ the API with I/O operations. So, I solved this case: 712127d
